File: python/fa_api.py
#!/usr/bin/env python
import urllib,simplejson,cgi

## Set some values we'll be needing through the process
CLIENT_ID="xxxx" # Issued by FormAssembly host
CLIENT_SECRET="xxxx" # Issued by FormAssembly host
RETURN_URL="https://localhost/" # We're doing commandline so this is pointless
AUTH_ENDPOINT="https://xxxxxx/oauth/login" # Replace xxxxxx with correct url
TOKEN_REQUEST_ENDPOINT="https://xxxxxx/oauth/access_token" # Replace xxxxxx with correct url

## API Endpoint to test
## https://server/api_v1/forms/index.xml
API_REQUEST="https://xxxxxx/api_v1/forms/index" 

## Build our authorization endpoint to display to user ('Adam')
AUTH_URI=AUTH_ENDPOINT+"?type=web&client_id="+CLIENT_ID+"&redirect_uri="+RETURN_URL+"&response_type=code"

## Since we're on the commandline, display authorization url to user ('Adam').
print "Go to URL: "+AUTH_URI
print "When directed to https://localhost/?code=XXXXXXXXXX copy and paste code here:\n"

code = raw_input("CODE>")
#Decode any URL encoded values
code = cgi.parse_qs("code="+code)["code"][0]

## We've got the code value, so we can now generate the server-side request for an access_token
TOKEN_REQUEST_DATA=urllib.urlencode({"grant_type":"authorization_code","type":"web_server","client_id":CLIENT_ID,"client_secret":CLIENT_SECRET,"redirect_uri":RETURN_URL,"code":code})

token_request_results=urllib.urlopen(TOKEN_REQUEST_ENDPOINT,TOKEN_REQUEST_DATA)
data = token_request_results.read()
print data

## Let's parse the response into an object from the json response, and grab the access_token
token=simplejson.loads(data)["access_token"]

format = raw_input("What format would you like the form list in? (json,xml,plist)> ")

## Make another server-side request for the API endpoint, using the access_token
FULL_API_REQUEST=API_REQUEST+"."+format+"?"+urllib.urlencode({"access_token":token})
api_response=urllib.urlopen(FULL_API_REQUEST)

## Display the output
print "\nURL:"+FULL_API_REQUEST+"\n"
for line in api_response:
  print line

File: node/fa_api.js
const request = require('request');
const readline = require('readline');
const querystring = require('querystring');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Set some values we'll be needing through the process
const CLIENT_ID="xxxx" // Issued by FormAssembly host
const CLIENT_SECRET="xxxx" // Issued by FormAssembly host
const RETURN_URL="https://localhost/" // We're doing commandline so this is pointless
const AUTH_ENDPOINT="https://xxxx/oauth/login" // Replace xxxxxx with correct url
const TOKEN_REQUEST_ENDPOINT="https://xxxx/oauth/access_token" // Replace xxxxxx with correct urlconst 

// API Endpoint to test
// https://server/api_v1/forms/index.xml
const API_REQUEST="https://xxxx/api_v1/forms/index"

// Build the query string
const API_AUTH_QUERY = {
  type:'web',
  client_id:CLIENT_ID,
  redirect_uri:RETURN_URL,
  response_type:'code'
};

let unescapedQS = querystring.unescape(querystring.stringify(API_AUTH_QUERY, '&', '=', { encode: false }))

// Build our authorization endpoint to display to user ('Adam')
const AUTH_URI=AUTH_ENDPOINT+"?"+unescapedQS;

// Since we're on the commandline, display authorization url to user ('Adam').
console.log("Go to URL: ", AUTH_URI);
console.log("When directed to https://localhost/?code=XXXXXXXXXX copy and paste code here: (do not include the ending #)\n");

rl.setPrompt('CODE> ');
rl.prompt();

rl.on('line', (input) => {
  let code = input;

  let qObj = {grant_type:'authorization_code',type:'web_server',client_id:CLIENT_ID,client_secret:CLIENT_SECRET,redirect_uri:RETURN_URL,code:code};

  let post_data = querystring.unescape(querystring.stringify(qObj));

  request({
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Content-Length': Buffer.byteLength(post_data),
      'Accept': 'application/json'
    },
    url: TOKEN_REQUEST_ENDPOINT, 
    method: 'POST', 
    body: post_data}, (error, response, body) => {
      if (error) {
        console.error(error);
      }
      console.log(response.statusCode);
      console.log(body);

      let token = JSON.parse(body)['access_token'];
      if (token != null) {
        rl.question('What format would you like the form list in? (json,xml,plist)> ', (format) => {

          const FULL_API_REQUEST=API_REQUEST+'.'+format+'?access_token='+token;
          console.log('Calling...', FULL_API_REQUEST);
          request({url: FULL_API_REQUEST, method: 'GET'}, (error, response, body) => {
            if (error) {
              console.error(error);
            }
            console.log(response.statusCode);
            console.log(body);
            rl.close();
          });
        });
      } else {
        rl.close();
      }
    });
});
